Add characterization tests for some "futils" function

piponazo opened the pull request:
In this PR I am adding characterization tests for some of the functions in **futils.hpp**. Since I want to keep the PRs small I decided to stop in the moment that I decided to change an implementation detail in **exiv2::urlencode**.

Here I am proposing to change the signature of exiv2::urlencode from

```
char* urlencode(char* str)
```
to
```
std::string urlencode(char* str)
```
The reasoning is to free client code from the responsibility of freeing the memory allocated by the method.

The function was only used in two places inside `src/basicio.cpp` and in the new added test.

piponazo committed:
add few characterisation tests for exiv2::strError

piponazo committed:
Ignore build* folders and project file generated by QtCreator

piponazo committed:
Add characterization tests for exiv2::getEnv

piponazo committed:
Add characterisation test for exiv2::urlencode

piponazo committed:
Change exiv2::urlencode signature to return std::string

The goal of this change is to remove the responsibility from the client code to
free the memory of the returned string.

piponazo committed:
Auto-formatting done by the IDE while touching code in the previous commits

D4N commented on src/futils.cpp:
```diff
@@ -85,8 +85,10 @@ namespace Exiv2 {
-        return buf;
-    } // urlencode
+        std::string ret(buf);
+        free(buf);
```
Good catch. Judging from your submitted diff, this was never freed' again. Congratulations for the fixed memory leak!

Could you change `malloc` & `free` to `new` & `delete[]`? The use of the C functions is discouraged in C++.

piponazo replied:
It was not a memory leak, the code in `src/basicio.cpp` was releasing correctly the memory, but since Exiv2 is a C++ project I thought it does not make sense to leave the memory management responsibility in the client code.

D4N commented:
The PR is segfaulting with clang on Travis during the `getEnv` test. Are you or gtest maybe implicitly assuming that glibc's `getenv` is thread safe (because it is absolutely not)?

piponazo commented:
I noticed the error on clang and I also have to addapt the code for windows :sweat_smile: . Fortunately, now we have the CI checks to tell us about these problems :)

piponazo committed:
Replace malloc/free by new/delete. Use const char* for input arg

piponazo committed:
exiv2::getEnv throws std::out_of_range on unexpected EnVar

This change define explicitly the behavior that exiv2::getEnv should have
on response to unexpected inputs.

There are some other minor changes:

- Use _putenv_s for the unit tests on Windows
- Add todo comment
- Remove deprecated note about freeing memory

D4N commented on src/futils.cpp:
```diff
@@ -58,9 +59,13 @@ namespace Exiv2 {
+    std::string getEnv(EnVar var)
+    {
+        if (var < envHTTPPOST || var > envTIMEOUT) {
+            throw std::range_error("Unexpected env variable");
```
I think most of the exiv2 code throws only instances of `Exiv2::Error`. To avoid bad surprises I would keep it that way rather than throwing an exception from std, although it might be better suited here.

piponazo replied:
True, I did not think about that. I will change it

piponazo replied:
Hummm, I have been checking this out and I do not like much the idea of using `Exiv2::Error` here. The constructor of this class always takes an integer, something that does not make sense here (Unless it already exists a table of values, and some of them could be used for cases like this one.) Otherwise I think that and std exception like `std::range_error` is better suited here.

clanmills replied:
I thought that @D4N was going to add an enum kerBlaDeeBla .... to give symbol names the error.  That's what we should be throwing and I believe they will be localised.

piponazo replied:
I understand but if that enum is not available yet ... what should I do ? Would be this enough?

```
if (var < envHTTPPOST || var > envTIMEOUT) {
    const int REPLACE_BY_ENUM_VALUE=123;
    throw Error(REPLACE_BY_ENUM_VALUE);
}
```
So that we can localise these values once the enum is ready ?

D4N replied:
I didn't have time to work on #109, so the enums are unfortunately not yet available.

How about we solve this issue in the following way: we replace the variables `ENVARDEF` and `ENVARKEY` with a `std::vector/array<std::string>`. Thereby, we do not need to perform a range check on the enum, as std::vector will do that itself (provided we use `.at()` in the `getEnv` function). Then throwing a `range_error` is imho ok here, as you really have to force it as a library user, provided that we document it.

piponazo replied:
This sounds like a perfect good solution to me :). Then, instead of throwing `range_error` , the std::vector::at() method throws with `out_of_range`:

http://www.cplusplus.com/reference/vector/vector/at/

I will commit the change in few minutes

D4N commented on src/futils.cpp:
```diff
@@ -71,9 +76,9 @@ namespace Exiv2 {
-        char* buf  = (char*)malloc(strlen(str) * 3 + 1);
+    std::string urlencode(const char* str) {
+        const char* pstr = str;
+        char* buf  = new char[strlen(str) * 3 + 1];
```
Depending on your level of motivation, you could change this to the `std::string` as creating it further below causes a copy of the whole string. Instead `buf` itself could be an instance of `std::string` constructed with the correct length.

piponazo replied:
Honestly, I am not very motivated to do it :P .

I changed the signature of the function because I saw some value on using more c++ types and removing from the client code the responsibility of freeing the C-strings. 

But I would not spend more time is this code unless it is strictly necessary. There are other more important things to do in the code base :)

D4N replied:
Fair enough. Can you add a `\todo` somewhere so that it is not forgotten?

piponazo commented on src/futils.cpp:
```diff
@@ -54,13 +55,15 @@ extern int strerror_r(int errnum, char *buf, size_t n);
 namespace Exiv2 {
-    const char* ENVARDEF[] = {"/exiv2.php", "40"}; //!< @brief default URL for http exiv2 handler and time-out
-    const char* ENVARKEY[] = {"EXIV2_HTTP_POST", "EXIV2_TIMEOUT"}; //!< @brief request keys for http exiv2 handler and time-out
+    std::vector<std::string> ENVARDEF = {"/exiv2.php", "40"};
```
Ahggg damn it! with c++98 we cannot do basic stuff like this ... Honestly, I would prefer to stay with the C-style arrays and throw the exception manually rather than add all the boiler plate needed to initialise a std::vector in C++98. 

Note: this compiled in Visual Studio 2017 since it is using all the power of modern c++ by default. There is no way to force it to use C++98. Recently they added some new flags to chose which standard to use ... but it starts from c++14:

https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/

In gcc and clang, this failed.

piponazo replied:
I will drop the latest commit and throw `std::out_of_range` manually. Whenever we change to modern c++, we could replace that piece of code easily.

D4N replied:
What the hell, you can't tell visual studio to use C++98? That sounds like a pretty crazy decision.

D4N commented on include/exiv2/futils.hpp:
```diff
@@ -78,7 +79,7 @@ namespace Exiv2 {
       @note Be sure to free() the returned string after use
```
This note is now obsolete, isn't it?

piponazo replied:
Right. We can remove that line

D4N reviewed (approved):
Once the outdated note is removed, I think we can merge this.

piponazo commented:
Cool, thanks for the useful feedback @D4N . I squashed some of the latest commits. I will wait for the CI checks to merge.
